### PR TITLE
Adding admin user AFTER enabling authentication

### LINKED ISSUE #535: Admin user creation in mongodb_auth fails because authentication is not enabled ?
##### SUMMARY

Here is my playbook content
```
---
- hosts: all

  become: true

  pre_tasks:

    - name: Install module PyMongo
      ansible.builtin.apt:
        name: python3-pymongo
        state: present

  roles:
    - community.mongodb.mongodb_linux
    - community.mongodb.mongodb_repository
    - community.mongodb.mongodb_install
    - community.mongodb.mongodb_mongod
    - community.mongodb.mongodb_auth

```

It fails at task `community.mongodb.mongodb_auth : Add mongo admin user with localhost exception` with `command createUser requires authentication, full error: {'ok': 0.0, 'errmsg': 'command createUser requires authentication', 'code': 13, 'codeName': 'Unauthorized'}`

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
community.mongodb.mongodb_auth

##### ANSIBLE VERSION
```
ansible [core 2.13.7]
  config file = <path-to-project>/ansible-mongodb-logs/ansible.cfg
  configured module search path = ['$HOME/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = $HOME/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
  jinja version = 3.0.3
  libyaml = True
```

##### COLLECTION VERSION

```
community.mongodb             1.4.2
```

##### CONFIGURATION
```
DEFAULT_ROLES_PATH('<path-to-project>/ansible-mongodb-logs/ansible.cfg) = ['<path-to-project>/ansible-mongodb-logs/roles']
HOST_KEY_CHECKING(<path-to-project>/ansible-mongodb-logs/ansible.cfg) = False
```

##### OS / ENVIRONMENT
Ubuntu 22.10 (local)
Debian 11 (remote)

##### STEPS TO REPRODUCE
Play the playbook printed above, without any variable override, it will fail. I personnally override some of them (mongodb_admin_pwd at least), but it fails the same way.

The task `Add mongo admin user with localhost exception` (https://github.com/ansible-collections/community.mongodb/blob/master/roles/mongodb_auth/tasks/main.yml#L38) uses `community.mongodb.mongodb_user` that does not work before authentication is enabled (it's done just after trying to add admin user).

##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_community.mongodb.mongodb_user_payload_tys0a5kq/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 389, in main
  File "/tmp/ansible_community.mongodb.mongodb_user_payload_tys0a5kq/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 266, in user_add
  File "/usr/lib/python3/dist-packages/pymongo/database.py", line 738, in command
    return self._command(sock_info, command, slave_ok, value,
  File "/usr/lib/python3/dist-packages/pymongo/database.py", line 626, in _command
    return sock_info.command(
  File "/usr/lib/python3/dist-packages/pymongo/pool.py", line 683, in command
    return command(self, dbname, spec, slave_ok,
  File "/usr/lib/python3/dist-packages/pymongo/network.py", line 159, in command
    helpers._check_command_response(
  File "/usr/lib/python3/dist-packages/pymongo/helpers.py", line 167, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response,
pymongo.errors.OperationFailure: command createUser requires authentication, full error: {'ok': 0.0, 'errmsg': 'command createUser requires authentication', 'code': 13, 'codeName': 'Unauthorized'}
fatal: [mongodb-log1]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "auth_mechanism": null,
            "connection_options": null,
            "create_for_localhost_exception": "/root/mongodb_admin.success",
            "database": "admin",
            "login_database": "admin",
            "login_host": "localhost",
            "login_password": null,
            "login_port": 27017,
            "login_user": null,
            "name": "admin",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "replica_set": null,
            "roles": [
                "root"
            ],
            "ssl": false,
            "ssl_ca_certs": null,
            "ssl_cert_reqs": "CERT_REQUIRED",
            "ssl_certfile": null,
            "ssl_crlfile": null,
            "ssl_keyfile": null,
            "ssl_pem_passphrase": null,
            "state": "present",
            "strict_compatibility": true,
            "update_password": "always"
        }
    },
    "msg": "Unable to add or update user: command createUser requires authentication, full error: {'ok': 0.0, 'errmsg': 'command createUser requires authentication', 'code': 13, 'codeName': 'Unauthorized'}"
}

```

### PR DESCRIPTION
**Pull request by alterphp**
Adding admin user through localhost exception is only allowed when security authentication is enabled, not before.

Fix #535

### CODE REVIEW

**Comment by rhysmeister** [comment]
@alterphp Thanks for the contribution. This certainly worked the old way, we have tests in https://github.com/ansible-collections/community.mongodb/tree/master/roles/mongodb_auth/molecule/default so I'm not sure why you had a problem before. The new way works anyway and if that's what the manual states I'm inclined to merge this PR. @cognifloyd Any thoughts?

Cheers,

Rhys

**Comment by alterphp** [comment]
Thanks for the feedback.

The test seems to test that default admin user password works after executing the playbook, which is fine if you do not try to change admin password. 

I'm gonna give a try to a test that changes default admin user to highlight a difference.

**Comment by rhysmeister** [comment]
The password is just what we use in the test. What I'm saying is that the order, of creating the admin user first, did work.

File: roles/mongodb_auth/molecule/default/tests/main.yml
Comment (rhysmeister) [inline roles/mongodb_auth/molecule/default/tests/main.yml:11]: This file should be put back to what it was. I have spotted the original problem.

File: roles/mongodb_auth/molecule/default/tests/main.yml
Comment (rhysmeister) [inline roles/mongodb_auth/molecule/default/tests/main.yml:13]: Note this variable name... **mongodb_default_admin_pwd**

File: roles/mongodb_auth/molecule/default/tests/test_default.py
Comment (rhysmeister) [inline roles/mongodb_auth/molecule/default/tests/test_default.py:47]: Note the variable name here. This lookup would have always failed and the default value would have been provided. THis line should be
`pwd = facts.get("mongodb_admin_pwd", "admin")`

**Comment by alterphp** [comment]
I've reverted the last commit. I was trying to test that admin password is well changed because it seems that the testsuite only tests that connection default admin password is OK. In practice, we should always change admin password (the steps that failed in my playbook).

**Review by cognifloyd — CHANGES_REQUESTED** [review]
No. I would rather not reorder these tasks.

Keep in mind that this should work on:
- a new cluster
- an existing cluster that you are adding auth to
- to validate that an existing cluster has auth enabled

For the last item, note the `create_for_localhost_exception` path. If that path is present, it should skip this task, assuming the admin was already created.

_edit: oh. I see the docs you linked to about Mango 5+_

**Comment by cognifloyd** [comment]
Will this still work with mongo 4 or 4.4? Or only 5+?

**Comment by rhysmeister** [comment]
> Will this still work with mongo 4 or 4.4? Or only 5+?

We only test on 5 & 6 now. As far as I know it'll still work. 

@alterphp - I can't actually ysee the problem you originally described. So somethign else must be going on I guess.

**Comment by alterphp** [comment]
It seems that the process has changed from MongoDB 4.4 to 5.0 :
- Doc for 4.4 explicitly tells to create the admin user first, then enable authentication : https://www.mongodb.com/docs/v4.4/tutorial/enable-authentication/#procedure
- Doc from 5.0 tells to enable authentication, then change admin user, thanks to the localhost exception (that did not exist before) : https://www.mongodb.com/docs/v5.0/core/localhost-exception/
I think the old way had a security disadvantage : a server without authentication might be "stolen" if someone could set the admin user password then enable authentication by gaining a local access.

The problem with the actual task order is that the task **Add mongo admin user with localhost exception** fails because authentication MUST BE enabled in order to change/create any user. The following tasks seems to be applied (I don't understand why), so authentication is enabled. The test checks that authentication is enabled by attempting a connection with admin user with **default** password : this does not prove that the task **Add mongo admin user with localhost exception** is a success, unless the testsuite tries to change the admin password => if attempting to connect admin user with the custom password successes, then it's the proof that localhost exception has been successfully used (and it covers what is considered as a highly recommended practice).

**Review by cognifloyd — APPROVED** [review]
Well, changing the order of the tasks means it won't support mongo 4 then. I need mongo 4 but oh well. That ship has sailed. I guess reordering the tasks is the right thing here.

**Comment by alterphp** [comment]
Fresh installation will more likely concern MongoDB 5+, and the roles will merely be used for maintenance with MongoDB 4-.

Ideally, it should order the tasks according to the mongodb server version, but I guess an highlighted note in the role could be sufficient.

**Comment by rhysmeister** [comment]
@cognifloyd  - Your call.
@alterphp - While 4.4 is still an acitvely maintained version it can be a pain supporting everthing. Users of 4.4 can still use an old version of the collection if needed. Soemtimes it's good to get rid of the bagge.

I've opened up two additional PR's related to this...

- https://github.com/ansible-collections/community.mongodb/pull/538
- https://github.com/ansible-collections/community.mongodb/pull/537

**Comment by rhysmeister** [comment]
Hello @alterphp, I think it's worth nothing here that your change does seem to work with MongoDB  4.4. The mongodb_repository role still had 4.4 for the mongodb_version variable in default/main.yml. I've updated this in the linked PR but the CI here ran using 4.4. I'm assuming it'll be good for 6.0 once merged. I'll merge if there's no further feedback.

**Comment by alterphp** [comment]
Thanks for the support of this library. I'm new to Ansible and its ecosystem, and I progressively discover the problems it solves, but also how hard it is to solve them !

**Pull request merged**
